Mathlib/Data/Pi/Algebra.lean

```diff
@@ -0,0 +1,398 @@
+  rfl
+
+-- -- Porting note : deprecated. not ported.
+-- @[simp]
+-- theorem bit0_apply [∀ i, Add <| f i] : (bit0 x) i = bit0 (x i) :=
+--   rfl
+--
+-- @[simp]
+-- theorem bit1_apply [∀ i, Add <| f i] [∀ i, One <| f i] : (bit1 x) i = bit1 (x i) :=
+--   rfl
+
+@[to_additive]
+instance instInv [∀ i, Inv <| f i] : Inv (∀ i : I, f i) :=
+  ⟨fun f i => (f i)⁻¹⟩
+
+#align pi.has_inv Pi.instInv
+
+@[simp, to_additive]
+theorem inv_apply [∀ i, Inv <| f i] : x⁻¹ i = (x i)⁻¹ :=
+  rfl
+
+@[to_additive]
+theorem inv_def [∀ i, Inv <| f i] : x⁻¹ = fun i => (x i)⁻¹ :=
+  rfl
+
+@[to_additive]
+theorem const_inv [Inv β] (a : β) : (const α a)⁻¹ = const α a⁻¹ :=
+  rfl
+
+@[to_additive]
+theorem inv_comp [Inv γ] (x : β → γ) (y : α → β) : x⁻¹ ∘ y = (x ∘ y)⁻¹ :=
+  rfl
+
+@[to_additive]
+instance instDiv [∀ i, Div <| f i] : Div (∀ i : I, f i) :=
+  ⟨fun f g i => f i / g i⟩
+
+#align pi.has_div Pi.instDiv
+
+@[simp, to_additive]
+theorem div_apply [∀ i, Div <| f i] : (x / y) i = x i / y i :=
+  rfl
+
+@[to_additive]
+theorem div_def [∀ i, Div <| f i] : x / y = fun i => x i / y i :=
+  rfl
+
+@[to_additive]
+theorem div_comp [Div γ] (x y : β → γ) (z : α → β) : (x / y) ∘ z = x ∘ z / y ∘ z :=
+  rfl
+
+@[simp, to_additive]
+theorem const_div [Div β] (a b : β) : const α a / const α b = const α (a / b) :=
+  rfl
+
+section
+
+variable [DecidableEq I]
+
+variable [∀ i, One (f i)] [∀ i, One (g i)] [∀ i, One (h i)]
+
+/-- The function supported at `i`, with value `x` there, and `1` elsewhere. -/
+@[to_additive Pi.Single "The function supported at `i`, with value `x` there, and `0` elsewhere."]
+def mulSingle (i : I) (x : f i) : ∀ (j : I), f j :=
+  Function.update 1 i x
+
+@[simp, to_additive]
+theorem mul_single_eq_same (i : I) (x : f i) : mulSingle i x i = x :=
+  Function.update_same i x _
+
+@[simp, to_additive]
+theorem mul_single_eq_of_ne {i i' : I} (h : i' ≠ i) (x : f i) : mulSingle i x i' = 1 :=
+  Function.update_noteq h x _
+
+/-- Abbreviation for `mul_single_eq_of_ne h.symm`, for ease of use by `simp`. -/
+@[simp, to_additive "Abbreviation for `single_eq_of_ne h.symm`, for ease of\nuse by `simp`."]
+theorem mul_single_eq_of_ne' {i i' : I} (h : i ≠ i') (x : f i) : mulSingle i x i' = 1 :=
+  mul_single_eq_of_ne h.symm x
+
+@[simp, to_additive]
+theorem mul_single_one (i : I) : mulSingle i (1 : f i) = 1 :=
+  Function.update_eq_self _ _
+
+-- Porting notes:
+-- 1) Why do I have to specify `(fun (_ : I) => β)` explicitely?
+-- 2) Why do I have to specify the type of `(1 : I → β)`?
+-- 3) Removed `{β : Sort _}` as `[One β]` converts it to a type anyways.
+/-- On non-dependent functions, `Pi.mulSingle` can be expressed as an `ite` -/
+@[to_additive "On non-dependent functions, `Pi.Single` can be expressed as an `ite`"]
+theorem mul_single_apply [One β] (i : I) (x : β) (i' : I) :
+    @mulSingle _ (fun (_ : I) => β) _ _ i x i' = if i' = i then x else 1 :=
+  Function.update_apply (1 : I → β) i x i'
```

Comment:
Lean3 found `(fun (_ : I) => β)` automatically, Lean4 seems not to. I'm not sure if it's bad to specify this now manually. 

Moreover without the type in `(1 : I → β)`, it complains that some universe is not bigger than zero.